docs/best-practices.rst

```diff
@@ -0,0 +1,271 @@
+    bucket_key="s3://bucket/key/foo.parquet"
+ )
+ task.set_downstream(check)
+
+
+
+Staging environment
+--------------------
+
+Always keep a staging environment to test the complete DAG run before deploying in the production.
+Make sure your DAG is parameterized to change the variables, e.g., the output path of S3 operation or the database used to read the configuration.
+Do not hard code values inside the DAG and then change them manually according to the environment.
+
+You can use Airflow Variables to parameterize the DAG.
+
+.. code::
+
+ dest = Variable(
+    "my_dag_dest",
+    "s3://default-target/path/"
+ )
+
+Deployment in Production
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Once you have completed all the mentioned checks, it is time to deploy your DAG in production.
+To do this, first, you need to make sure that the Airflow is itself production-ready. 
+Let's see what precautions you need to take.
+
+
+Backend
+--------
+
+Airflow comes with an ``SQLite`` backend by default. It allows the user to run Airflow without any external database.
+However, such a setup is meant to be for testing purposes only. Running the default setup can lead to data loss in multiple scenarios. 
+If you want to run Airflow in production, make sure you :doc:`configure the backend <howto/initialize-database>` to be an external database such as ``MySQL`` or ``Postgres``. 
+
+You can change the backend using the following config-
+
+.. code::
+
+ [core]
+ sql_alchemy_conn = my_conn_string
+
+Once you have changed the backend, airflow needs to create all the tables required for operation.
+Create an empty DB and give airflow's user the permission to ``CREATE/ALTER`` it.
+Once that is done, you can run -
+
+.. code::
+
+ airflow upgradedb
+
+``upgradedb`` keeps track of migrations already applies, so it's safe to run as often as you need.
+
+.. note::
+ 
+ Do not use ``airflow initdb`` as it can create a lot of default connection, charts, etc. which are not required in production DB.
+
+
+Multi-Node Cluster
+-------------------
+
+Airflow uses :class:`airflow.executors.sequential_executor.SequentialExecutor` by default. It works fine in most cases. However, by its nature, the user is limited to executing at most
+one task at a time. It's also not suitable to work in a multi-node cluster. You should use :doc:`../executor/celery` or :doc:`../executor/kubernetes` in such cases.
+
+
+Once you have configured the executor, it is necessary to make sure that every node in the cluster contains the same configuration and dags.
+Airflow only sends simple instructions such as execute task X on node Y but does not send any dag files or configuration. You can use a simple CRON or
+any other mechanism to sync DAGs and configs across your nodes, e.g., checkout DAGs from git repo every 5 minutes on all nodes.
+
+
+Logging
+--------
+
+If you are using disposable nodes in your cluster, configure the log storage to be a distributed file system such as ``S3`` or ``GCS``.
```

Comment:
Airflow can also be used with other tools that do not provide a file system e.g. Stackdriver Logging (not publicly available yet), Elasticsearch, Amazon CloudWatch.

Reply:
Agreed.